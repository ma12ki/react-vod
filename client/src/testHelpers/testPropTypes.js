import React from 'react';

// this has some limitations:
// https://stackoverflow.com/questions/41916992/why-does-my-jest-test-of-react-proptypes-break-when-using-multiple-unacceptable/41916993#41916993
export const testPropTypes = (component, propName, acceptableValues, unacceptableValues, otherProps) => {
    console.error = jest.fn();
    const _test = (testValues, expectError) => {
        for (let propValue of testValues) {
            console.error.mockClear();
            React.createElement(component, {...otherProps, [propName]: propValue});
            try {
                expect(console.error).toHaveBeenCalledTimes(expectError ? 1 : 0);
            } catch (err) {
                console.log(`${propName} expected to ${expectError ? '' : 'NOT'} fail for value ${propValue}`);
                throw err;
            }
        }
    };
    _test(acceptableValues || [], false);
    _test(unacceptableValues || [], true);
};
