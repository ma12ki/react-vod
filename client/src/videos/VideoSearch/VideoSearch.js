import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Input } from 'reactstrap';
import { debounce } from 'throttle-debounce';

import { getSearch } from '../selectors';
import { search } from '../ducks';

export class VideoSearch extends React.PureComponent {
    static propTypes = {
        search: PropTypes.string.isRequired,
        performSearch: PropTypes.func.isRequired,
    }

    debouncedSearch = debounce(250, false, (term) => {
        this.props.performSearch(term);
    });

    render() {
        const { search } = this.props;

        return <Input defaultValue={search} onKeyUp={(e) => this.debouncedSearch(e.target.value)} placeholder='search files...' />;
    }
}

const mapStateToProps = (state) => ({
    search: getSearch(state),
});

const mapDispatchToProps = (dispatch) => ({
    performSearch: (term) => dispatch(search(term)),
});

export default connect(mapStateToProps, mapDispatchToProps)(VideoSearch);
