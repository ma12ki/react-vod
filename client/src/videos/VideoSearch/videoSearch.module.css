.input.input {
    border-radius: 0;
}

.input.input:global(::placeholder) {
    opacity: 0.5;
}
