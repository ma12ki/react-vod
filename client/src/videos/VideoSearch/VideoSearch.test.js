import React from 'react';
import { shallow, mount } from 'enzyme';

import { testPropTypes } from '../../testHelpers';
import { VideoSearch } from './VideoSearch';

jest.mock('throttle-debounce', () => {
    return {
        debounce: (_delay, _atBegin, fn) => (...args) => fn.apply(null, args)
    };
});

const makeWrapper = ({ search = '', performSearch = () => null } = {}) => shallow(<VideoSearch search={search} performSearch={performSearch} />);
const mountWrapper = ({ search = '', performSearch = () => null } = {}) => mount(<VideoSearch search={search} performSearch={performSearch} />);

describe('VideoSearch', () => {

    beforeEach(() => jest.clearAllMocks());

    it('matches the snapshot', () => {
        const wrapper = makeWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    it('requires "search" string prop', () => {
        testPropTypes(
            VideoSearch,
            'search',
            ['rofl', 'xD', ''],
            [123, null, undefined],
            {
                performSearch: () => null
            }
        );
    });

    it('requires "performSearch" function prop', () => {
        testPropTypes(
            VideoSearch,
            'performSearch',
            [() => null],
            ['123', 123, null, undefined],
            {
                search: 'lol'
            }
        );
    });

    it('calls "performSearch" on key up', () => {
        const performSearch = jest.fn();
        const wrapper = mountWrapper({ performSearch });
        const chars = 'lol';

        wrapper.find('input').node.value = chars;
        wrapper.simulate('keyup');

        expect(performSearch).toHaveBeenCalledWith(chars);
    });

});