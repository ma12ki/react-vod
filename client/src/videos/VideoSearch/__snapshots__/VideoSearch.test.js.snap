// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoSearch matches the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Input
    defaultValue=""
    onKeyUp={[Function]}
    placeholder="search files..."
    tag="p"
    type="text"
/>,
  "nodes": Array [
    <Input
      defaultValue=""
      onKeyUp={[Function]}
      placeholder="search files..."
      tag="p"
      type="text"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <VideoSearch
        performSearch={[Function]}
        search=""
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": VideoSearch {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "debouncedSearch": [Function],
        "props": Object {
          "performSearch": [Function],
          "search": "",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Input
          defaultValue=""
          onKeyUp={[Function]}
          placeholder="search files..."
          tag="p"
          type="text"
/>,
        "_debugID": 2,
        "_renderedOutput": <Input
          defaultValue=""
          onKeyUp={[Function]}
          placeholder="search files..."
          tag="p"
          type="text"
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <VideoSearch
    performSearch={[Function]}
    search=""
/>,
}
`;
